external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import scala.util.Try
+
+import kafka.common.TopicAndPartition
+import org.apache.kafka.clients.producer.RecordMetadata
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.sql.StreamTest
+import org.apache.spark.sql.execution.streaming.{Offset, Source}
+import org.apache.spark.sql.streaming.OffsetSuite
+import org.apache.spark.sql.test.SharedSQLContext
+
+
+class KafkaSourceSuite extends StreamTest with SharedSQLContext with OffsetSuite {
+
+  import testImplicits._
+
+  private var testUtils: KafkaTestUtils = _
+  private var kafkaParams: Map[String, String] = _
+
+  override val streamingTimout = 10.seconds
+
+  case class AddKafkaData(
+      kafkaSource: KafkaSource,
+      topic: String, data: Int*)(implicit multiPartitionCheck: Boolean = true) extends AddData {
+
+    override def addData(): Offset = {
+      val sentMetadata = testUtils.sendMessages(topic, data.map{ _.toString}.toArray)
+      val latestOffsetMap = sentMetadata
+        .groupBy { _._2.partition }
+        .map { case (partition, msgAndMetadata) =>
+          val maxOffsetInPartition = msgAndMetadata.map(_._2).maxBy(_.offset).offset()
+          (TopicAndPartition(topic, partition), maxOffsetInPartition + 1)
+        }
+
+      def metadataToStr(m: (String, RecordMetadata)): String = {
+        s"Sent ${m._1} to partition ${m._2.partition()}, offset ${m._2.offset()}"
+      }
+
+      assert(latestOffsetMap.size > 1,
+        s"Added data does not test multiple partitions: " + sentMetadata.map(metadataToStr))
```

Comment:
This is to make sure that data is added such that it gets added to multiple Kafka partitions. This will throw an exception if some one does not give data such that it gets spread across partitions.

Just to make the test future proof, so that someone accidentally does not disable multi-partition testing. Probably better to make this a `require` instead of an `assert`.